<a href="https://colab.research.google.com/github/nananana25/MSC151/blob/main/Bitcoin_test.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import yfinance as yf

tickers = ["BTC-USD", "GC=F", "^DJI", "CL=F"]
start_date = "2025-01-01"
end_date = "2025-06-30"

data = yf.download(tickers, start=start_date, end=end_date)

print(data.head())

/tmp/ipython-input-3080639487.py:7: FutureWarning: YF.download() has changed argument auto_adjust default to True
  data = yf.download(tickers, start=start_date, end=end_date)
[*********************100%***********************]  4 of 4 completed

Price              Close                                                High  \
Ticker           BTC-USD       CL=F         GC=F          ^DJI       BTC-USD   
Date                                                                           
2025-01-01  94419.757812        NaN          NaN           NaN  94929.867188   
2025-01-02  96886.875000  73.129997  2658.899902  42392.269531  97739.820312   
2025-01-03  98107.429688  73.959999  2645.000000  42732.128906  98956.914062   
2025-01-04  98236.226562        NaN          NaN           NaN  98734.429688   
2025-01-05  98314.960938        NaN          NaN           NaN  98813.304688   

Price                                                      Low             \
Ticker           CL=F         GC=F          ^DJI       BTC-USD       CL=F   
Date                                                                        
2025-01-01        NaN          NaN           NaN  92788.125000        NaN   
2025-01-02  73.730003  2663.100098  42905.089844  9

# Task
Extract the 'Close' prices from the `data` DataFrame, drop any rows with missing values, calculate the daily returns for each asset, and then rename the columns to 'bitcoin', 'gold', 'US', and 'oil' respectively.

## Extract Close prices and drop NaNs

### Subtask:
Select the 'Close' prices from the multi-indexed DataFrame and remove any rows containing NaN values from these selected columns.


**Reasoning**:
I will extract the 'Close' prices from the multi-indexed `data` DataFrame and then remove any rows with NaN values, storing the result in a new DataFrame called `close_prices`.



In [2]:
close_prices = data['Close'].dropna()
print(close_prices.head())

Ticker            BTC-USD       CL=F         GC=F          ^DJI
Date                                                           
2025-01-02   96886.875000  73.129997  2658.899902  42392.269531
2025-01-03   98107.429688  73.959999  2645.000000  42732.128906
2025-01-06  102078.085938  73.559998  2638.399902  42706.558594
2025-01-07   96922.703125  74.250000  2656.699951  42528.359375
2025-01-08   95043.523438  73.320000  2664.500000  42635.199219


## Calculate daily returns

### Subtask:
Compute the daily percentage change (returns) for each asset based on their 'Close' prices.


**Reasoning**:
I will calculate the daily percentage change for each column in the `close_prices` DataFrame using the `.pct_change()` method and store it in `daily_returns`. Then, I will display the first few rows of the `daily_returns` DataFrame.



In [3]:
daily_returns = close_prices.pct_change()
print(daily_returns.head())

Ticker       BTC-USD      CL=F      GC=F      ^DJI
Date                                              
2025-01-02       NaN       NaN       NaN       NaN
2025-01-03  0.012598  0.011350 -0.005228  0.008017
2025-01-06  0.040473 -0.005408 -0.002495 -0.000598
2025-01-07 -0.050504  0.009380  0.006936 -0.004173
2025-01-08 -0.019388 -0.012525  0.002936  0.002512


## Rename columns

### Subtask:
Rename the columns of the returns DataFrame to more descriptive names: 'BTC-USD' to 'bitcoin', 'CL=F' to 'gold', '^DJI' to 'US', and 'GC=F' to 'oil'.


## Summary:

### Q&A
The task asked to extract 'Close' prices, drop missing values, calculate daily returns, and then rename the columns. The provided solving process successfully extracted 'Close' prices, dropped missing values, and calculated daily returns. However, the renaming of columns to 'bitcoin', 'gold', 'US', and 'oil' was not included in the presented steps.

### Data Analysis Key Findings
*   The 'Close' prices for various assets (`BTC-USD`, `CL=F`, `GC=F`, `^DJI`) were successfully extracted from the `data` DataFrame and all rows containing missing values were removed, resulting in a clean `close_prices` DataFrame.
*   Daily percentage changes (returns) were accurately computed for each asset based on their 'Close' prices.
*   The `daily_returns` DataFrame correctly shows `NaN` values for the first row, as there's no prior data point for calculation.
*   For example, on '2025-01-03', Bitcoin's price increased by approximately 1.26%, while Gold experienced a decrease of approximately 0.52%.

### Insights or Next Steps
*   The next step in the overall task would be to rename the columns of the `daily_returns` DataFrame from their current symbols to more descriptive names like 'bitcoin', 'gold', 'US', and 'oil' for better readability and interpretability.
*   The `daily_returns` DataFrame is now prepared for further financial analysis, such as volatility calculations, correlation analysis, or portfolio optimization.


# Task
Rename the columns of the `daily_returns` DataFrame from 'BTC-USD' to 'bitcoin', 'CL=F' to 'gold', '^DJI' to 'US', and 'GC=F' to 'oil'.

## Rename columns

### Subtask:
Rename the columns of the `daily_returns` DataFrame from 'BTC-USD' to 'bitcoin', 'CL=F' to 'gold', '^DJI' to 'US', and 'GC=F' to 'oil'.


**Reasoning**:
I will rename the columns of the `daily_returns` DataFrame using a mapping dictionary as specified in the instructions and then display the head of the DataFrame to verify the changes.



In [4]:
column_mapping = {'BTC-USD': 'bitcoin', 'GC=F': 'gold', '^DJI': 'US', 'CL=F': 'oil'}
daily_returns.rename(columns=column_mapping, inplace=True)
print(daily_returns.head())

Ticker       bitcoin       oil      gold        US
Date                                              
2025-01-02       NaN       NaN       NaN       NaN
2025-01-03  0.012598  0.011350 -0.005228  0.008017
2025-01-06  0.040473 -0.005408 -0.002495 -0.000598
2025-01-07 -0.050504  0.009380  0.006936 -0.004173
2025-01-08 -0.019388 -0.012525  0.002936  0.002512


**Reasoning**:
The previous renaming step incorrectly mapped 'CL=F' to 'oil' and 'GC=F' to 'gold', instead of 'CL=F' to 'gold' and 'GC=F' to 'oil' as specified in the subtask. To correct this, I will first regenerate the `daily_returns` DataFrame from `close_prices` to restore the original column names, and then apply the correct column mapping.



In [6]:
daily_returns = close_prices.pct_change()

column_mapping = {'BTC-USD': 'bitcoin', 'CL=F': 'gold', 'GC=F': 'oil', '^DJI': 'US'}
daily_returns.rename(columns=column_mapping, inplace=True)
print(daily_returns.head())

Ticker       bitcoin      gold       oil        US
Date                                              
2025-01-02       NaN       NaN       NaN       NaN
2025-01-03  0.012598  0.011350 -0.005228  0.008017
2025-01-06  0.040473 -0.005408 -0.002495 -0.000598
2025-01-07 -0.050504  0.009380  0.006936 -0.004173
2025-01-08 -0.019388 -0.012525  0.002936  0.002512


## Run Regression Model

### Subtask:
Prepare the data by selecting 'bitcoin' as the dependent variable and 'gold', 'oil', and 'US' as independent variables, add an intercept, and then run an Ordinary Least Squares (OLS) regression.

**Reasoning**:
I will first import the `statsmodels.api` library. Then, I'll define the dependent variable `y` as 'bitcoin' and the independent variables `X` as 'gold', 'oil', and 'US' from the `daily_returns` DataFrame. I will add a constant to `X` to include an intercept in the model. Finally, I will fit an OLS regression model and print its summary to analyze the results.

In [7]:
import statsmodels.api as sm

# Drop NaN values from daily_returns for regression
daily_returns_cleaned = daily_returns.dropna()

# Define dependent variable (y)
y = daily_returns_cleaned['bitcoin']

# Define independent variables (X)
X = daily_returns_cleaned[['gold', 'oil', 'US']]

# Add a constant (intercept) to the independent variables
X = sm.add_constant(X)

# Create and fit the OLS model
model = sm.OLS(y, X)
results = model.fit()

# Print the regression summary
print(results.summary())

                            OLS Regression Results                            
Dep. Variable:                bitcoin   R-squared:                       0.120
Model:                            OLS   Adj. R-squared:                  0.098
Method:                 Least Squares   F-statistic:                     5.287
Date:                Tue, 18 Nov 2025   Prob (F-statistic):            0.00188
Time:                        09:10:16   Log-Likelihood:                 271.30
No. Observations:                 120   AIC:                            -534.6
Df Residuals:                     116   BIC:                            -523.4
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          0.0009      0.002      0.360      0.7